README.md
# 概述

提供grpc-kit微服务脚手架的通用配置规范

## 术语约定

### 连接地址

参数 | 类型 | 说明 | 示例
-----|------|---------|----
host | string | 主机IPv4或IPv6地址 | 127.0.0.1
port | int | 主机端口号 | 6379
address | string | 不包含协议，由主机IP与端口组成 | 127.0.0.1:6379 或 [fe80::1%lo0]:53
endpoints | []string | 由协议、IP、端口组成，多个通过逗号","分割 | https://node1:2379,https://node2:2379,https://node3:2379

### 其他类型

参数 | 类型 | 说明 | 示例
-----|------|---------|----
driver | string | 配置支持的驱动 | etcdv3、redis等
enable | bool | 是否开启这个功能 | true、false

## 配置示例

```yaml
# https://github.com/grpc-kit/cfg/blob/master/app-sample.yaml

# 基础服务配置
services:
  # 服务注册的前缀，全局统一
  root_path: service
  # 服务注册的空间，全局统一
  namespace: example
  # 服务的代码，设置后不可变更
  service_code: cmdb.v1.commons
  # 接口网关的地址
  api_endpoint: api.grpc-kit.com
  # 服务所监听的grpc地址（如未设置，自动监听在127.0.0.1的随机端口）
  grpc_address: 127.0.0.1:10081
  # 服务所监听的http地址（如未设置，则不开启gateway服务）
  http_address: 127.0.0.1:10080
  # 服务注册，外部网络可连接的grpc地址（一般等同于grpc-address）
  public_address: ""

# 服务注册配置
discover:
  driver: etcdv3
  heartbeat: 15
  endpoints:
    - http://127.0.0.1:2379
  #discover:
  #  tls:
  #    ca_file: /opt/certs/etcd-ca.pem
  #    cert_file: /opt/certs/etcd.pem
  #    key_file: /opt/certs/etcd-key.pem

# 认证鉴权配置
security:
  enable: true
  # 认证：谁在登录
  authentication:
    # 跳过认证的rpc方法
    insecure_rpcs:
      - SearchHosts
    oidc_provider:
      issuer: https://accounts.example.com
      config:
        # 必须验证token.aud是否与client_id相等
        client_id: example
        # 允许的签名算法类别
        supported_signing_algs:
          - RS256
        # 忽略token.aud与client_id的验证
        skip_client_id_check: true
        # 忽略token是否过期的验证
        skip_expiry_check: false
        # 忽略token issuer的验证
        skip_issuer_check: true
    http_users:
      - username: user1
        password: pass1
  # TODO; 鉴权：能做什么
  authorization:

# 关系数据配置
database:
  driver: postgres
  dbname: demo
  user: demo
  password: grpc-kit
  host: 127.0.0.1
  port: 5432
  sslmode: disable
  connect_timeout: 10

# 缓存服务配置
cachebuf:
  enable: true
  driver: redis
  address: 127.0.0.1:6379
  password: ""

# 日志调试配置
debugger:
  enable_pprof: true
  log_level: debug
  log_format: text

# 链路追踪配置
opentracing:
  enable: true
  host: 127.0.0.1
  port: 6831

# 应用私有配置
independent:
  name: grpc-kit
```

### 注册地址说明

在services中所配置的public-address地址是grpc服务，与其他服务之间必须能正常通讯。如果在k8s环境下可考虑设置环境变量GRPC_KIT_PUHLIC_IP把POD IP传递，如：

```yaml
...
spec:
  template:
    spec:
      containers:
      - args:
        - /opt/service
        - --config
        - /opt/config/app.toml
        env:
        - name: GRPC_KIT_PUHLIC_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.podIP
...
```


CHANGELOG.md
# Changelog

名称 | 说明
------------|----------
Added       | 添加新功能
Changed     | 功能的变更
Deprecated  | 未来会删除
Removed     | 之前为Deprecated状态，此版本被移除
Fixed       | 功能的修复
Security    | 有关安全问题的修复

## [Unreleased]

## [0.1.3] - 2021-06-02

### Added

- 更改配置toml为yaml格式
- 添加http basic与oidc认证支持

## [0.1.2] - 2021-01-07

### Changed

- 根配置结构更改为引用形式，防止默认类型的零值，困扰判断类型存在
- 服务注册内容不在提供用户自定义，强制写入本地配置内容至注册中心
- 用户可自定义gRPC的一元拦截器，并以自定义为高优先级可覆盖默认值
- 简化模版注册流程，对gateway的handler也通过cfg.Register方法实现

### Added

- 通用gRPC部分方法的实现
- 实现服务注册与撤销方法
- 服务注册配置新增Heartbeat时间，代表ttl
- 客户端与服务端默认流拦截器的支持
- 支持GRPC_KIT_PUHLIC_IP变量实现IP地址注册
- 可支持配置pprof的开关闭

## [0.1.0] - 2020-02-14

### Added

- 首次发布


cfg_test.go
package cfg

import (
	"fmt"
	"math/rand"
	"testing"
	"time"

	"github.com/opentracing/opentracing-go"
	"github.com/spf13/viper"
)

var lc *LocalConfig

func TestConfig(t *testing.T) {
	var err error

	viper.SetConfigType("toml")
	viper.SetConfigFile("app-sample.toml")

	if err := viper.ReadInConfig(); err != nil {
		t.Errorf("Load config file err: %v\n", err)
	}

	lc, err = New(viper.GetViper())
	if err != nil {
		t.Errorf("Load config file err: %v\n", err)
	}

	if lc.Services.Namespace != "example" {
		t.Errorf("service.Namespace != example")
	}

	t.Run("testServiceName", testServiceName)
	t.Run("testInitLogger", testLogger)
	t.Run("testInitOpenTracing", testOpenTracing)
	// t.Run("testServiceGRPCAddress", testServiceGRPCAddress)
}

func testServiceName(t *testing.T) {
	serviceName := "cmdb.v1.commons.api.grpc-kit.com"
	if serviceName != fmt.Sprintf("%v.%v", lc.Services.ServiceCode, lc.Services.APIEndpoint) {
		t.Errorf("service_name not match")
	}
}

func testLogger(t *testing.T) {
	lo, err := lc.InitLogger()
	if err != nil {
		t.Errorf("Init logger err: %v\n", err)
	}

	lo.Info("hello test")
}

func testOpenTracing(t *testing.T) {
	closer, err := lc.InitOpenTracing()
	if err != nil {
		t.Errorf("Init OpenTracing err: %v\n", err)
	}
	defer closer.Close()

	rand.Seed(time.Now().UnixNano())

	tracer := opentracing.GlobalTracer()
	rootSpan := tracer.StartSpan("Hello_RootSpan")

	for i := 0; i <= 2; i++ {
		var csp opentracing.Span
		if i == 1 {
			csp = opentracing.StartSpan(fmt.Sprintf("FollowsFrom_%v", i), opentracing.FollowsFrom(rootSpan.Context()))
		} else {
			csp = opentracing.StartSpan(fmt.Sprintf("ChildOf_%v", i), opentracing.ChildOf(rootSpan.Context()))
		}

		randInt := rand.Intn(100)
		csp.SetTag("process_time", randInt)
		time.Sleep(time.Duration(randInt) * time.Millisecond)

		for j := 0; j <= 1; j++ {

			var fsp opentracing.Span
			if j == 1 {
				fsp = opentracing.StartSpan(fmt.Sprintf("FollowsFrom_%v", j), opentracing.FollowsFrom(csp.Context()))
			} else {
				fsp = opentracing.StartSpan(fmt.Sprintf("ChildOf_%v", j), opentracing.ChildOf(csp.Context()))
			}

			randInt := rand.Intn(100)
			fsp.SetTag("process_time", randInt)
			time.Sleep(time.Duration(randInt) * time.Millisecond)

			fsp.SetTag("follows_key", fmt.Sprintf("%v", j))
			fsp.Finish()

			go func() {
				for x := 0; x <= 1; x++ {
					var xsp opentracing.Span
					if x == 1 {
						xsp = opentracing.StartSpan(fmt.Sprintf("FollowsFrom_%v", x), opentracing.FollowsFrom(fsp.Context()))
					} else {
						xsp = opentracing.StartSpan(fmt.Sprintf("ChildOf_%v", x), opentracing.ChildOf(fsp.Context()))
					}

					randInt := rand.Intn(100)
					xsp.SetTag("process_time", randInt)
					time.Sleep(time.Duration(randInt) * time.Millisecond)

					xsp.Finish()
				}
			}()
		}

		csp.Finish()
	}

	rootSpan.Finish()

	time.Sleep(3 * time.Second)
}

/*
func testServiceGRPCAddress(t *testing.T) {
	address, port, err := lc.Services.GetGRPCListenHostPort()
	if err != nil {
		t.Errorf("err: %v", err)
	}

	fmt.Println("grpc-host:", address, "grpc-port:", port)
}
*/


auth.go
package cfg

import (
	"context"
	"errors"
	"time"

	"github.com/coreos/go-oidc/v3/oidc"
	"github.com/dgrijalva/jwt-go"
	"k8s.io/apimachinery/pkg/util/wait"
)

// IDTokenClaims 用于框架jwt的数据结构
type IDTokenClaims struct {
	jwt.StandardClaims
	Email           string            `json:"email"`
	EmailVerified   bool              `json:"email_verified"`
	Groups          []string          `json:"groups"`
	FederatedClaims map[string]string `json:"federated_claims"`
}

// InitAuthentication 初始化认证
func (c *LocalConfig) InitAuthentication() error {
	if !c.Security.Enable {
		return nil
	}

	if c.Security.Authentication == nil {
		return errors.New("security authentication enable but unset")
	}

	// 初始化jwt token认证
	if c.Security.Authentication.OIDCProvider != nil {
		if c.Security.Authentication.OIDCProvider.Issuer == "" {
			return errors.New("security authentication not found oidc issuer")
		}

		ctx := context.TODO()

		initVerifierFn := func() (done bool, err error) {
			provider, err := oidc.NewProvider(ctx, c.Security.Authentication.OIDCProvider.Issuer)
			if err != nil {
				c.logger.Errorf("oidc authenticator: initializing plugin: %v", err)
				return false, err
			}

			oidcConfig := &oidc.Config{}
			if c.Security.Authentication.OIDCProvider.Config != nil {
				oidcConfig.ClientID = c.Security.Authentication.OIDCProvider.Config.ClientID
				oidcConfig.SupportedSigningAlgs = c.Security.Authentication.OIDCProvider.Config.SupportedSigningAlgs
				oidcConfig.SkipClientIDCheck = c.Security.Authentication.OIDCProvider.Config.SkipClientIDCheck
				oidcConfig.SkipExpiryCheck = c.Security.Authentication.OIDCProvider.Config.SkipExpiryCheck
				oidcConfig.SkipIssuerCheck = c.Security.Authentication.OIDCProvider.Config.SkipIssuerCheck
			}

			verifier := provider.Verifier(oidcConfig)
			c.Security.setVerifier(verifier)

			return true, nil
		}

		ok, err := initVerifierFn()
		if !ok || err != nil {
			go wait.PollUntil(time.Second*30, initVerifierFn, ctx.Done())
		}
	}

	return nil
}

// WithIDToken 用于设置当前会话的IDToken
func (c *LocalConfig) WithIDToken(parent context.Context, token IDTokenClaims) context.Context {
	return context.WithValue(parent, idTokenKey, token)
}

// WithUsername 用于设置当前会话的用户名
func (c *LocalConfig) WithUsername(parent context.Context, username string) context.Context {
	return context.WithValue(parent, usernameKey, username)
}

// WithAuthenticationType 用于设置当前会话的认证方式
func (c *LocalConfig) WithAuthenticationType(parent context.Context, authType string) context.Context {
	return context.WithValue(parent, authenticationTypeKey, authType)
}

// IDTokenFrom 用于获取当前会话的IDToken
func (c *LocalConfig) IDTokenFrom(ctx context.Context) (IDTokenClaims, bool) {
	idToken, ok := ctx.Value(idTokenKey).(IDTokenClaims)
	return idToken, ok
}

// UsernameFrom 用于获取当前会话的用户名
func (c *LocalConfig) UsernameFrom(ctx context.Context) (string, bool) {
	username, ok := ctx.Value(usernameKey).(string)
	return username, ok
}

// AuthenticationTypeFrom 用于获取当前会话的认证方式
func (c *LocalConfig) AuthenticationTypeFrom(ctx context.Context) (string, bool) {
	username, ok := ctx.Value(authenticationTypeKey).(string)
	return username, ok
}

// setVerifier 用于设置oidc verifier实例
func (s *SecurityConfig) setVerifier(v *oidc.IDTokenVerifier) {
	s.tokenVerifier.Store(v)
}

// idTokenVerifier 用于获取odic verifier实例
func (s *SecurityConfig) idTokenVerifier() (*oidc.IDTokenVerifier, bool) {
	if v := s.tokenVerifier.Load(); v != nil {
		return v.(*oidc.IDTokenVerifier), true
	}
	return nil, false
}


logger.go
package cfg

import (
	"os"

	"github.com/sirupsen/logrus"
)

// InitLogger 用于初始化日志实例
func (c *LocalConfig) InitLogger() (*logrus.Entry, error) {
	logger := logrus.WithFields(
		logrus.Fields{
			"service_name": c.GetServiceName(),
		})

	logLevel := c.Debugger.LogLevel
	logFormat := c.Debugger.LogFormat

	if logLevel == "" {
		logLevel = "error"
	}
	if logFormat == "" {
		logFormat = "text"
	}

	switch logLevel {
	case "panic":
		logrus.SetLevel(logrus.PanicLevel)
	case "fatal":
		logrus.SetLevel(logrus.FatalLevel)
	case "error":
		logrus.SetLevel(logrus.ErrorLevel)
	case "warn":
		logrus.SetLevel(logrus.WarnLevel)
	case "info":
		logrus.SetLevel(logrus.InfoLevel)
	case "debug":
		logrus.SetLevel(logrus.DebugLevel)
	default:
		logrus.SetLevel(logrus.WarnLevel)
	}

	switch logFormat {
	case "json":
		logrus.SetFormatter(&logrus.JSONFormatter{})
	case "text":
		logrus.SetFormatter(&logrus.TextFormatter{
			DisableColors: true,
		})
	default:
		logrus.SetFormatter(&logrus.TextFormatter{})
	}

	logrus.SetOutput(os.Stdout)

	c.logger = logger

	return logger, nil
}

// GetLogger 用于获取全局日志
func (c *LocalConfig) GetLogger() *logrus.Entry {
    return c.logger
}

opentracing.go
package cfg

import (
	"fmt"
	"io"
	// "time"

	"github.com/uber/jaeger-client-go"
	jaegerconfig "github.com/uber/jaeger-client-go/config"
)

// InitOpenTracing 初始化全局分布式链路追踪
func (c *LocalConfig) InitOpenTracing() (io.Closer, error) {
	samplerCfg := &jaegerconfig.SamplerConfig{
		Type:  jaeger.SamplerTypeConst,
		Param: 1,
	}

	if c.Opentracing.Host == "" {
		c.Opentracing.Host = "127.0.0.1"
	}
	if c.Opentracing.Port == 0 {
		c.Opentracing.Port = 6831
	}
	tracingHost := fmt.Sprintf("%v:%v", c.Opentracing.Host, c.Opentracing.Port)

	reporterCfg := &jaegerconfig.ReporterConfig{
		// QueueSize: 1,
		// LogSpans:            false,
		// BufferFlushInterval: 3 * time.Second,
		LocalAgentHostPort: tracingHost,
	}

	headerCfg := &jaeger.HeadersConfig{
		TraceContextHeaderName:   "jaeger-trace-id",
		TraceBaggageHeaderPrefix: "jaeger-ctx",
	}

	jaegerCfg := jaegerconfig.Configuration{
		Disabled: !c.Opentracing.Enable,
		Sampler:  samplerCfg,
		Reporter: reporterCfg,
		Headers:  headerCfg}

	closer, err := jaegerCfg.InitGlobalTracer(c.GetServiceName())
	if err != nil {
		return closer, err
	}

	return closer, nil
}


cfg.go
package cfg

import (
	"context"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"os"
	"strings"
	"sync/atomic"
	"time"

	"github.com/grpc-ecosystem/grpc-gateway/runtime"
	"github.com/grpc-kit/pkg/sd"
	"github.com/mitchellh/mapstructure"
	"github.com/sirupsen/logrus"
	"github.com/spf13/viper"
	"google.golang.org/grpc"
	"google.golang.org/grpc/resolver"
)

const (
	// HTTPHeaderRequestID 全局请求ID
	HTTPHeaderRequestID = "X-TR-REQUEST-ID"
	// TraceContextHeaderName 链路追踪ID
	TraceContextHeaderName = "jaeger-trace-id"
	// TraceBaggageHeaderPrefix 数据传递头前缀
	TraceBaggageHeaderPrefix = "jaeger-ctx"
	// AuthenticationTypeBasic 用于http basic认证
	AuthenticationTypeBasic = "basic"
	// AuthenticationTypeBearer 用于jwt认证
	AuthenticationTypeBearer = "bearer"
	// AuthenticationTypeNone 用于指明rpc未使用任何认证
	AuthenticationTypeNone = "none"
	// UsernameAnonymous 当未使用任何认证时的用户名
	UsernameAnonymous = "anonymous"
)

// contextKey 使用自定义类型不对外，防止碰撞冲突
type contextKey int

const (
	// idTokenKey 用于存放当前jwt的解析后的数据结构
	idTokenKey contextKey = iota

	// usernameKey 用于存放当前用户名，http base对应username，jwt对应email
	usernameKey

	// authenticationTypeKey 用于存放当前认证方式
	authenticationTypeKey
)

// LocalConfig 本地配置，全局微服务配置结构
type LocalConfig struct {
	Services    *ServicesConfig    `json:",omitempty"` // 基础服务配置
	Discover    *DiscoverConfig    `json:",omitempty"` // 服务注册配置
	Security    *SecurityConfig    `json:",omitempty"` // 认证鉴权配置
	Database    *DatabaseConfig    `json:",omitempty"` // 关系数据配置
	Cachebuf    *CachebufConfig    `json:",omitempty"` // 缓存服务配置
	Debugger    *DebuggerConfig    `json:",omitempty"` // 日志调试配置
	Opentracing *OpentracingConfig `json:",omitempty"` // 链路追踪配置
	Independent interface{}        `json:",omitempty"` // 应用私有配置

	logger *logrus.Entry
	srvdis sd.Clienter
}

// ServicesConfig 基础服务配置，用于设定命名空间、注册的路径、监听的地址等
type ServicesConfig struct {
	RootPath      string `mapstructure:"root_path"`
	Namespace     string `mapstructure:"namespace"`
	ServiceCode   string `mapstructure:"service_code"`
	APIEndpoint   string `mapstructure:"api_endpoint"`
	GRPCAddress   string `mapstructure:"grpc_address"`
	HTTPAddress   string `mapstructure:"http_address"`
	PublicAddress string `mapstructure:"public_address"`
}

// DiscoverConfig 服务注册，服务启动后如何汇报自身
type DiscoverConfig struct {
	Driver    string     `mapstructure:"driver"`
	Endpoints []string   `mapstructure:"endpoints"`
	TLS       *TLSConfig `mapstructure:"tls" json:",omitempty"`
	Heartbeat int64      `mapstructure:"heartbeat"`
}

// SecurityConfig 安全配置，对接认证、鉴权
type SecurityConfig struct {
	// 包含*oidc.IDTokenVerifier数据结构，不能直接使用*oidc.IDTokenVerifier
	tokenVerifier atomic.Value

	Enable         bool            `mapstructure:"enable"`
	Authentication *Authentication `mapstructure:"authentication"`
	Authorization  *Authorization  `mapstructure:"authorization"`
}

// DatabaseConfig 数据库设置，指关系数据库，数据不允许丢失，如postgres、mysql
type DatabaseConfig struct {
	Driver         string `mapstructure:"driver"`
	DBName         string `mapstructure:"dbname"`
	User           string `mapstructure:"user"`
	Password       string `mapstructure:"password"`
	Host           string `mapstructure:"host"`
	Port           int    `mapstructure:"port"`
	SSLMode        string `mapstructure:"sslmode"`
	ConnectTimeout int    `mapstructure:"connect_timeout"`
}

// CachebufConfig 缓存配置，区别于数据库配置，缓存的数据可以丢失
type CachebufConfig struct {
	Enable   bool   `mapstructure:"enable"`
	Driver   string `mapstructure:"driver"`
	Address  string `mapstructure:"address"`
	Password string `mapstructure:"password"`
}

// DebuggerConfig 日志配置，用于设定服务启动后日志输出级别格式等
type DebuggerConfig struct {
	EnablePprof bool   `mapstructure:"enable_pprof"`
	LogLevel    string `mapstructure:"log_level"`
	LogFormat   string `mapstructure:"log_format"`
}

// OpentracingConfig 分布式链路追踪
type OpentracingConfig struct {
	Enable bool   `mapstructure:"enable"`
	Host   string `mapstructure:"host"`
	Port   int    `mapstructure:"port"`
}

// TLSConfig 证书配置
type TLSConfig struct {
	CAFile             string `mapstructure:"ca_file"`
	CertFile           string `mapstructure:"cert_file"`
	KeyFile            string `mapstructure:"key_file"`
	InsecureSkipVerify bool   `mapstructure:"insecure_skip_verify"`
}

// Authentication 用于认证
type Authentication struct {
	InsecureRPCs []string      `mapstructure:"insecure_rpcs"`
	OIDCProvider *OIDCProvider `mapstructure:"oidc_provider"`
	HTTPUsers    []*BasicAuth  `mapstructure:"http_users"`
}

// Authorization 用于鉴权
type Authorization struct {
}

// BasicAuth 用于HTTP基本认证的用户权限定义
type BasicAuth struct {
	Username string `mapstructure:"username"`
	Password string `mapstructure:"password"`
}

// OIDCProvider 用于OIDC认证提供方配置
type OIDCProvider struct {
	Issuer string      `mapstructure:"issuer"`
	Config *OIDCConfig `mapstructure:"config"`
}

// OIDCConfig 用于OIDC验证相关配置
type OIDCConfig struct {
	ClientID             string   `mapstructure:"client_id"`
	SupportedSigningAlgs []string `mapstructure:"supported_signing_algs"`
	SkipClientIDCheck    bool     `mapstructure:"skip_client_id_check"`
	SkipExpiryCheck      bool     `mapstructure:"skip_expiry_check"`
	SkipIssuerCheck      bool     `mapstructure:"skip_issuer_check"`
}

// New 用于初始化获取全局配置实例
func New(v *viper.Viper) (*LocalConfig, error) {
	var lc LocalConfig

	if err := viper.Unmarshal(&lc); err != nil {
		return nil, err
	}

	// 验证几个关键属性是否在
	if lc.Services.RootPath == "" {
		return nil, fmt.Errorf("unknow root_path")
	}
	if lc.Services.Namespace == "" {
		return nil, fmt.Errorf("unknow namespace")
	}
	if lc.Services.ServiceCode == "" {
		return nil, fmt.Errorf("unknow service_code")
	}
	if lc.Services.APIEndpoint == "" {
		return nil, fmt.Errorf("unknow api_endpoint")
	}

	// 初始化默认设置
	if lc.Services.GRPCAddress == "" {
		rand.Seed(time.Now().UnixNano())
		lc.Services.GRPCAddress = fmt.Sprintf("127.0.0.1:%v", 10081+rand.Intn(6000))
	}
	if lc.Services.PublicAddress == "" {
		// 支持环境变量设置微服务地址
		if addr := os.Getenv("GRPC_KIT_PUHLIC_IP"); addr != "" {
			// 获取服务端口
			tmp := strings.Split(lc.Services.GRPCAddress, ":")
			if len(tmp) == 2 {
				lc.Services.PublicAddress = fmt.Sprintf("%v:%v", addr, tmp[1])
			} else {
				lc.Services.PublicAddress = lc.Services.GRPCAddress
			}
		} else {
			lc.Services.PublicAddress = lc.Services.GRPCAddress
		}
	}

	return &lc, nil
}

// Init 用于根据配置初始化各个实例
func (c *LocalConfig) Init() error {
	if _, err := c.InitLogger(); err != nil {
		return err
	}

	if _, err := c.InitOpenTracing(); err != nil {
		return err
	}

	if err := c.InitAuthentication(); err != nil {
		return err
	}

	return nil
}

// Register 用于登记服务信息至注册中心
func (c *LocalConfig) Register(ctx context.Context,
	gw func(context.Context, *runtime.ServeMux, string, []grpc.DialOption) (err error),
	opts ...runtime.ServeMuxOption) (*http.ServeMux, error) {

	if err := c.registerConfig(); err != nil {
		return nil, err
	}

	return c.registerGateway(ctx, gw, opts...)
}

// Deregister 用于撤销注册中心上的服务信息
func (c *LocalConfig) Deregister() error {
	// 配置文件未设置注册地址，则主动忽略
	if c.Discover == nil {
		return nil
	}

	return c.srvdis.Deregister()
}

// GetIndependent 用于获取各个微服务独立的配置
func (c *LocalConfig) GetIndependent(t interface{}) error {
	if c.Independent == nil {
		return fmt.Errorf("independent is nil")
	}

	return mapstructure.Decode(c.Independent, t)
}

// GetServiceName 用于获取微服务名称
func (c *LocalConfig) GetServiceName() string {
	return fmt.Sprintf("%v.%v", c.Services.ServiceCode, c.Services.APIEndpoint)
}

func (c *LocalConfig) registerConfig() error {
	// 配置文件未设置注册地址，则主动忽略
	if c.Discover == nil {
		return nil
	}

	sd.Home(c.Services.RootPath, c.Services.Namespace)
	connector, err := sd.NewConnector(c.logger, sd.ETCDV3, strings.Join(c.Discover.Endpoints, ","))
	if err != nil {
		return err
	}

	if c.Discover.TLS != nil {
		tls := &sd.TLSInfo{
			CAFile:   c.Discover.TLS.CAFile,
			CertFile: c.Discover.TLS.CertFile,
			KeyFile:  c.Discover.TLS.KeyFile,
		}
		connector.WithTLSInfo(tls)
	}

	ttl := c.Discover.Heartbeat
	if ttl == 0 {
		ttl = 30
	}

	rawBody, err := json.Marshal(c)
	if err != nil {
		return err
	}

	reg, err := sd.Register(connector, c.GetServiceName(), c.Services.PublicAddress, string(rawBody), ttl)
	if err != nil {
		return fmt.Errorf("Register server err: %v%v", err, "\n")
	}

	c.srvdis = reg

	// 注册解析器
	r := sd.NewResolver(connector)
	resolver.Register(r)

	return nil
}
